keps/draft-20190516-KUDO-package-manager.md

```diff
@@ -0,0 +1,177 @@
+
+Overall, this KEP should capture how we plan to provide a great user experience of installing Frameworks with KUDO. Attention should be split to:
+
+* How the Package structure looks like and dictates the overall repository structure
+* What tools should be used to achieve this structure
+* How are packages are distributed to the user
+
+## Motivation
+
+There are multiple package managers and ideas out there and we need to decide which concept will work the best for KUDO in its current state. This is not an Install CLI subcategory KEP, but of course we need to think about the repercussions of decisions made in this KEP. There are multiple stakeholders directly being impacted by decisions made here, but this also paves the way to think about new problems we haven't articulated yet (e.g. how to provide KUDO Frameworks in air-gapped clusters, verification of Frameworks and their tarbundles, etc.).
+
+### Goals
+
+Specific goals are:
+* Coming up with a structure for a Package
+* Defining the overall `kubebuilder/framework` repository structure
+* Having a solution on how to distribute this structure
+* Having this platform-agnostic so Packages can be distributed across multiple solutions
+* Adapting concepts from other successfull open-source projects in this regard
+
+### Non-Goals
+
+Non-goals are:
+* Re-inventing the wheel
+* Solving problems that are not related to the overall goals
+* Some more that I can't think of yet
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is. The folder structure and what could be interpreted as our v0 of the Registry was originally discussed in PR [#87](https://github.com/kudobuilder/kudo/pull/87) and is a great way to start when thinking of this enhancement.
+
+### User Stories
+
+#### Framework Developer
+
+As a Framework Developer I, ...
+
+* would like to be able to host validation/testing plans and resources as part of the package
+* would like to update my package version
+
+#### Cluster Administrator
+
+As a Cluster Administrator I, ...
+
+* would like to be able to host my custom Framework extensions on a private KUDO Framework repo
+* would like to be able to add additional validation/testing plans and resources
+
+### Implementation Details/Notes/Constraints
+
+We need to make sure that Packages can be easily accessed even if the underlying Storage Backend changed. Therefor, adding a middle layer like a registry server that has multiple Storage Backends as a swappable engine underneath seems like a good pattern. This also would enable users to keep track or restrict access to specific Frameworks.
```

Comment:
I still struggle with the fact that our KEPs are just too huge for me. I think introducing package registry and ability to download package from any HTTP source or local file should be fine for now. We could maybe deal with air-gapped env later? Theoretically with what we do, air-gapped clusters are kind of supported already, it's just they would have to write the HTTP layer by themselves. By providing them the HTTP layer and multiple backends, we are just making it easier for them :)

Reply:
Yes. We can consider explicit airgap out of the initial scope.

Reply:
totally agree! 👍Scratching entire line 85 seems appropriate to me.